hx711_rpio.go

```diff
@@ -0,0 +1,129 @@
+// +build !windows,rpio
+
+package hx711
+
+import (
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/stianeikeland/go-rpio/v4"
+)
+
+// HostInit opens /dev/gpiomem. This needs to be done before Hx711 can be used.
+func HostInit() error {
+	return rpio.Open()
+}
+
+// NewHx711 creates new Hx711.
+// Make sure to set clockPinNum and dataPinNum to the correct pins.
+// The pin numbers must comply with BCM numbering schema.
+// https://cdn.sparkfun.com/datasheets/Sensors/ForceFlex/hx711_english.pdf
+// https://godoc.org/github.com/stianeikeland/go-rpio#Pin
+func NewHx711(clockPinNum string, dataPinNum string) (*Hx711, error) {
+	clockPin, err := strconv.ParseInt(clockPinStr, 10, 32)
+	if err != nil {
+		return nil, err
+	}
+	dataPin, err := strconv.ParseInt(dataPinStr, 10, 32)
+	if err != nil {
+		return nil, err
+	}
+	hx711 := &Hx711{numEndPulses: 1}
+	hx711.clockPin = rpio.Pin(int(clockPin))
+	hx711.dataPin = rpio.Pin(int(dataPin))
+	hx711.dataPin.Input()
+	hx711.clockPin.Output()
+	return hx711, nil
+}
+
+// setClockHighThenLow sets clock pin high then low
+func (hx711 *Hx711) setClockHighThenLow() error {
+	hx711.clockPin.Write(rpio.High)
+	hx711.clockPin.Write(rpio.Low)
+	return nil
+}
+
+// Reset starts up or resets the chip.
+// The chip needs to be reset if it is not used for just about any amount of time.
+func (hx711 *Hx711) Reset() error {
+	hx711.clockPin.Write(rpio.Low)
+	hx711.clockPin.Write(rpio.High)
+	time.Sleep(70 * time.Microsecond)
+	hx711.clockPin.Write(rpio.Low)
+	return nil
+}
+
+// Shutdown puts the chip in powered down mode.
+// The chip should be shutdown if it is not used for just about any amount of time.
+func (hx711 *Hx711) Shutdown() error {
+	hx711.clockPin.Write(rpio.High)
+	return nil
+}
+
+// waitForDataReady waits for data to go to low which means chip is ready
+func (hx711 *Hx711) waitForDataReady() error {
+	var level rpio.State
+
+	// looks like chip often takes 80 to 100 milliseconds to get ready
+	// but somettimes it takes around 500 milliseconds to get ready
+	// WaitForEdge sometimes returns right away
+	// So will loop for N times, which could be more than 1 second, but usually 500 milliseconds
+
+	hx711.dataPin.Detect(rpio.FallEdge)
+	hx711.clockPin.Write(rpio.Low)
+	defer hx711.dataPin.Detect(rpio.NoEdge)
+	level = hx711.dataPin.Read()
+	if level == rpio.Low {
+		return nil
+	}
+	for i := 0; i < 200000; i++ {
+		if hx711.dataPin.EdgeDetected() {
+			return nil
+		}
+		time.Sleep(5 * time.Microsecond)
```

Comment:
From what I found (see note in the code) it takes around 80 milliseconds for the chip to get ready. Did you find something different?

Still wondering if 5 microsecond sleep is to small.

Reply:
I was testing with different timeouts, my findings are:
- anything above 1 millisecond leads to unstable reading results ( about 5 - 10% errors )
- with 2 microseconds the number of iterations before the edge is detected is 27000-31000, which gives 27-30 milliseconds delay

I don't really know *why* setting up the delay of 1 millisecond causes errors. My suspicion is that 10 milliseconds is not really 1 millisecond.

The most stable and reasonable results that I've got were with 250 microseconds to wait until trying to read GPIO memory to see whether there was falling edge detected. The number of iterations is less than 500, but I feel like it could be a good idea to cap it to 1500, just in case.

Hope that helps :)

Reply:
Thank you so much, that is great :)